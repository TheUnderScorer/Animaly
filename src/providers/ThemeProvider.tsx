import React, {
  createContext,
  FC,
  useCallback,
  useContext,
  useState,
} from 'react';
import { ApplicationProvider, IconRegistry } from '@ui-kitten/components';
import * as eva from '@eva-design/eva';
import { EvaIconsPack } from '@ui-kitten/eva-icons';

const defaultValue: ThemeContextType = {
  theme: 'light',
  toggleTheme: () => {},
};

export interface ThemeContextType {
  theme: 'light' | 'dark';
  toggleTheme: () => void;
}

const ThemeContext = createContext<ThemeContextType>(defaultValue);
export const useThemeContext = () => useContext(ThemeContext);

const ThemeProvider: FC = ({ children }) => {
  const [theme, setTheme] = useState<'light' | 'dark'>('light');
  const toggleTheme = useCallback(() => {
    setTheme(theme === 'light' ? 'dark' : 'light');
  }, [theme]);

  return (
    <ThemeContext.Provider
      value={{
        theme,
        toggleTheme,
      }}>
      <IconRegistry icons={EvaIconsPack} />
      <ApplicationProvider {...eva} theme={eva[theme]}>
        {children}
      </ApplicationProvider>
    </ThemeContext.Provider>
  );
};

export default ThemeProvider;
