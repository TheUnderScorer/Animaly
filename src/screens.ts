export enum RootScreens {
  Home = 'Home',
  Splash = 'Splash',
}

export enum SplashScreens {
  Start = 'Start',
  CreateUserForm = 'CreateUserForm',
}

export enum AppScreens {
  Welcome = 'Welcome',
  Home = 'Home',
}
