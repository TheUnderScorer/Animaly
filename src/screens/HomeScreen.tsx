import React, { FC } from 'react';
import UserWelcome from '../ui/molecules/UserWelcome';
import useTimeOfDayBackgroundAndText from '../hooks/common/useTimeOfDayBackgroundAndText';
import { SafeAreaView, StatusBar, StyleSheet, View } from 'react-native';
import FeaturesGrid from '../ui/molecules/FeaturesGrid';
import { Layout } from '@ui-kitten/components';
import HomeBackground from '../ui/atoms/HomeBackground';

export interface HomeScreenProps {}

const styles = StyleSheet.create({
  layout: {
    height: '100%',
  },
  text: {
    color: '#fff',
  },
  inner: {
    bottom: 60,
  },
});

const HomeScreen: FC<HomeScreenProps> = () => {
  const { text, background } = useTimeOfDayBackgroundAndText();

  return (
    <Layout style={styles.layout} level="2" testID="homeScreen">
      <StatusBar
        barStyle="light-content"
        translucent
        backgroundColor="transparent"
      />
      <HomeBackground background={background}>
        <UserWelcome textStyle={styles.text} timeOfDay={text} />
      </HomeBackground>
      <SafeAreaView style={styles.layout}>
        <View style={styles.inner}>
          <FeaturesGrid />
        </View>
      </SafeAreaView>
    </Layout>
  );
};

export default HomeScreen;
