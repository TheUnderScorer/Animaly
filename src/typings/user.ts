export interface User {
  name: string;
  createdAt: Date;
}
