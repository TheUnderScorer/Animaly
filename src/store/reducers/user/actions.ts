import { Dispatch } from 'redux';
import { AppActions } from '../../index';
import AsyncStore from '../../../storage/AsyncStore';
import { User } from '../../../typings/user';
import { CreateUserInput } from './types';

export const fetchCurrentUser = (asyncStore: AsyncStore) => async (
  dispatch: Dispatch<AppActions>,
) => {
  const currentUser = await asyncStore.get<User>('currentUser');

  dispatch({
    type: 'SetCurrentUser',
    payload: currentUser,
  });

  return dispatch({
    type: 'SetDidInitialFetch',
    payload: true,
  });
};

export const createUser = (
  asyncStore: AsyncStore,
  input: CreateUserInput,
) => async (dispatch: Dispatch<AppActions>) => {
  const user: User = {
    ...input,
    createdAt: new Date(),
    hasBeenWelcomed: false,
  };

  asyncStore.set('currentUser', user);

  dispatch({
    type: 'SetCurrentUser',
    payload: user,
  });

  return user;
};
