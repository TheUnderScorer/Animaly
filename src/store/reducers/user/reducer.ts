import { User } from '../../../typings/user';
import { Action, defaultHandler, mapReducer } from '@theunderscorer/utils';

export interface UserState {
  currentUser: User | null;
  didInitialFetch: boolean;
}

export const initialState: UserState = {
  currentUser: null,
  didInitialFetch: false,
};

export type UserActions =
  | Action<'SetCurrentUser', User | null>
  | Action<'SetDidInitialFetch', boolean>;

export const userReducer = mapReducer<UserState, UserActions>(
  {
    SetCurrentUser: defaultHandler('currentUser'),
    SetDidInitialFetch: defaultHandler('didInitialFetch'),
  },
  initialState,
);
