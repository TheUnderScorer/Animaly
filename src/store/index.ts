import { applyMiddleware, combineReducers, createStore } from 'redux';
import { UserActions, userReducer, UserState } from './reducers/user/reducer';
import thunk from 'redux-thunk';
import { initialState as userInitialState } from './reducers/user/reducer';

export interface AppStore {
  user: UserState;
}

const initialState: AppStore = {
  user: userInitialState,
};

export type AppActions = UserActions;

const reducers = combineReducers({
  user: userReducer,
});
const middleware = [thunk];

const store = createStore(
  reducers,
  initialState,
  applyMiddleware(...middleware),
);

export default store;
