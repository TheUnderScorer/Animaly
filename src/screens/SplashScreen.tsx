import React, {
  FC,
  MutableRefObject,
  useCallback,
  useEffect,
  useRef,
  useState,
} from 'react';
import { Layout, Spinner } from '@ui-kitten/components';
import { RowView } from '../styles/view';
import {
  Image,
  StatusBar,
  StyleSheet,
  View,
  Text,
  Linking,
} from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import { AppStore } from '../store';
import {
  Transition,
  Transitioning,
  TransitioningView,
} from 'react-native-reanimated';
import { useAsyncStorageContext } from '../providers/AsyncStorageProvider';
import { fetchCurrentUser } from '../store/reducers/user/actions';
import AppLogo from '../ui/molecules/AppLogo';
import { useNavigation } from '@react-navigation/native';
import { RootScreens } from '../screens';
import { wait } from '../utils/timeout';
import CreateUserForm from '../ui/organisms/CreateUserForm';
import GetStartedBtn from '../ui/atoms/GetStartedBtn';
import Bg from '../assets/paws.jpg';
import { BackgroundOverlay } from '../styles/overlay';
import { A } from '../styles/typography';

export interface SplashScreenProps {}

const styles = StyleSheet.create({
  logo: {
    marginBottom: 40,
  },
  row: {
    justifyContent: 'center',
    alignItems: 'center',
    width: 300,
  },
  layout: {
    width: '100%',
    height: '100%',
    alignItems: 'center',
    justifyContent: 'center',
  },
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  bg: {
    position: 'absolute',
    width: '100%',
    height: '100%',
    zIndex: 0,
    top: 0,
  },
  bgSource: {
    position: 'absolute',
    color: '#fff',
    bottom: 20,
    left: 10,
  },
});

export type CreationState = 'notCreating' | 'creating';

const SplashScreen: FC<SplashScreenProps> = () => {
  const transitionRef = useRef<TransitioningView>();

  const [creationState, setCreationState] = useState<CreationState>(
    'notCreating',
  );

  const navigation = useNavigation();

  const { storage } = useAsyncStorageContext();

  const dispatch = useDispatch();
  const currentUser = useSelector<AppStore>((store) => store.user.currentUser);
  const didInitialUserFetch = useSelector<AppStore>(
    (store) => store.user.didInitialFetch,
  );
  const [loading, setLoading] = useState(false);

  const handleToggleIsCreatingUser = useCallback(() => {
    transitionRef.current!.animateNextTransition();
    setCreationState('creating');
  }, [transitionRef]);

  useEffect(() => {
    if (didInitialUserFetch || loading || !transitionRef.current) {
      return;
    }

    const timeout = setTimeout(async () => {
      transitionRef.current!.animateNextTransition();

      setLoading(true);

      await wait(1000);

      await dispatch(fetchCurrentUser(storage));

      setLoading(false);
    }, 1000);

    return () => {
      clearTimeout(timeout);
    };
  }, [
    didInitialUserFetch,
    currentUser,
    dispatch,
    loading,
    storage,
    transitionRef,
  ]);

  useEffect(() => {
    if (didInitialUserFetch) {
      if (currentUser) {
        navigation.navigate(RootScreens.Home);
      }
    }
  }, [didInitialUserFetch, currentUser, navigation, transitionRef]);

  return (
    <Transitioning.View
      style={styles.container}
      ref={transitionRef as MutableRefObject<TransitioningView>}
      transition={
        <Transition.Sequence>
          <Transition.Out type="fade" />
          <Transition.In type="fade" delayMs={500} />
          <Transition.Change interpolation="easeInOut" />
        </Transition.Sequence>
      }>
      <Layout level="4" style={styles.layout}>
        {creationState === 'notCreating' && (
          <>
            <Image source={Bg} style={styles.bg} />
            <BackgroundOverlay />
            <Text style={styles.bgSource}>
              Designed by{' '}
              <A
                variant="dark"
                onPress={() =>
                  Linking.openURL('https://www.freepik.com/pch-vector')
                }>
                pch-vector
              </A>
            </Text>
          </>
        )}
        <StatusBar hidden />
        {creationState === 'creating' && <CreateUserForm />}
        <View>
          {creationState === 'notCreating' && (
            <View style={styles.logo}>
              <AppLogo />
            </View>
          )}
          <RowView style={[styles.row]}>
            {loading && <Spinner testID="splashSpinner" size="giant" />}
            {didInitialUserFetch &&
              !currentUser &&
              creationState === 'notCreating' && (
                <GetStartedBtn onPress={handleToggleIsCreatingUser} />
              )}
          </RowView>
        </View>
      </Layout>
    </Transitioning.View>
  );
};

export default SplashScreen;
