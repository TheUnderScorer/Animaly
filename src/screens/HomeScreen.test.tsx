import React, { ReactNode } from 'react';
import { Provider } from 'react-redux';
import { makeStore } from '../store';
import AsyncStorageProvider from '../providers/AsyncStorageProvider';
import ThemeProvider from '../providers/ThemeProvider';
import { NavigationContainer } from '@react-navigation/native';
import { RootScreens } from '../screens';
import HomeScreen from './HomeScreen';
import { createMockProxy } from 'jest-mock-proxy/lib';
import AsyncStore from '../storage/AsyncStore';
import { act, render } from 'react-native-testing-library';
import { getMockUser } from '../tests/mocks/user';
import { wait } from '../utils/timeout';

const user = getMockUser();

const mockStore = createMockProxy<AsyncStore>();

const wrapComponent = (cmp: ReactNode) => (
  <Provider store={makeStore()}>
    <AsyncStorageProvider defaultStorage={mockStore}>
      <ThemeProvider>
        <NavigationContainer>{cmp}</NavigationContainer>
      </ThemeProvider>
    </AsyncStorageProvider>
  </Provider>
);

describe('<HomeScreen />', () => {
  beforeEach(() => {
    mockStore.mockClear();

    mockStore.get.mockResolvedValueOnce(user);
  });

  it('should render without crashing', async () => {
    const cmp = render(wrapComponent(<HomeScreen />));

    await act(async () => {
      await wait(100);
    });

    expect(cmp.toJSON()).toMatchSnapshot();
  });
});
