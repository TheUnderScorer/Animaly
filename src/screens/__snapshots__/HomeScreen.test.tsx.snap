// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomeScreen /> should render without crashing 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    appearance="default"
    level="2"
    style={
      Array [
        Object {
          "backgroundColor": "#F7F9FC",
        },
        Object {
          "height": "100%",
        },
      ]
    }
    testID="homeScreen"
  >
    <View
      accessibilityIgnoresInvertColors={true}
      style={
        Array [
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "maxHeight": 400,
            "overflow": "hidden",
            "paddingBottom": 40,
          },
          Object {
            "height": 400.2,
          },
        ]
      }
    >
      <Image
        source={
          Object {
            "uri": "https://source.unsplash.com/weekly?afternoon",
          }
        }
        style={
          Array [
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            Object {
              "height": undefined,
              "width": undefined,
            },
            undefined,
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "backgroundColor": "rgba(0, 0, 0, 0.3)",
              "height": "100%",
              "position": "absolute",
              "width": "100%",
              "zIndex": 0,
            },
            Object {
              "height": 1334,
            },
          ]
        }
      />
    </View>
    <RCTSafeAreaView
      emulateUnlessSupported={true}
      style={
        Object {
          "height": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "bottom": 60,
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
              "flexWrap": "wrap",
              "justifyContent": "space-evenly",
            }
          }
        >
          <View
            accessible={true}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#FFFFFF",
                "borderColor": "#E4E9F2",
                "borderRadius": 4,
                "borderWidth": 1,
                "height": 150,
                "justifyContent": "center",
                "minWidth": 150,
                "opacity": 1,
                "overflow": "hidden",
                "width": "45%",
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": undefined,
                  "height": 0,
                }
              }
            />
            <View
              style={
                Object {
                  "paddingHorizontal": 24,
                  "paddingVertical": 16,
                }
              }
            >
              <View
                accessible={true}
                focusable={false}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "justifyContent": "center",
                    "opacity": 1,
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": undefined,
                        "fontSize": 50,
                      },
                      undefined,
                      Object {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  
                </Text>
                <Text
                  appearance="default"
                  category="p1"
                  style={
                    Array [
                      Object {
                        "color": "#222B45",
                        "fontFamily": "System",
                        "fontSize": 15,
                        "fontWeight": "400",
                      },
                      Object {
                        "marginTop": 15,
                      },
                    ]
                  }
                >
                  Mind Clear
                </Text>
              </View>
            </View>
          </View>
          <View
            accessible={true}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#FFFFFF",
                "borderColor": "#E4E9F2",
                "borderRadius": 4,
                "borderWidth": 1,
                "height": 150,
                "justifyContent": "center",
                "minWidth": 150,
                "opacity": 1,
                "overflow": "hidden",
                "width": "45%",
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": undefined,
                  "height": 0,
                }
              }
            />
            <View
              style={
                Object {
                  "paddingHorizontal": 24,
                  "paddingVertical": 16,
                }
              }
            >
              <View
                accessible={true}
                focusable={false}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "justifyContent": "center",
                    "opacity": 1,
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": undefined,
                        "fontSize": 50,
                      },
                      undefined,
                      Object {
                        "fontFamily": "FontAwesome5Free-Solid",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {
                        "fontWeight": "900",
                      },
                    ]
                  }
                >
                  
                </Text>
                <Text
                  appearance="default"
                  category="p1"
                  style={
                    Array [
                      Object {
                        "color": "#222B45",
                        "fontFamily": "System",
                        "fontSize": 15,
                        "fontWeight": "400",
                      },
                      Object {
                        "marginTop": 15,
                      },
                    ]
                  }
                >
                  Animal Photos
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTSafeAreaView>
  </View>
</View>
`;
